roles/test-utils/mining-device/src/main.rs

```diff
@@ -499,3 +610,43 @@ impl Miner {
         }
     }
 }
+
+// returns hashrate based on how fast the device hashes over the given duration
+fn measure_hashrate(duration_secs: u64) -> f64 {
+    let mut rng = thread_rng();
+    let prev_hash: [u8; 32] = generate_random_32_byte_array().to_vec().try_into().unwrap();
+    let prev_hash = Hash::from_inner(prev_hash);
+    let merkle_root: [u8; 32] = generate_random_32_byte_array().to_vec().try_into().unwrap();
+    let merkle_root = Hash::from_inner(merkle_root);
+    let header = BlockHeader {
+        version: rng.gen(),
+        prev_blockhash: BlockHash::from_hash(prev_hash),
+        merkle_root,
+        time: std::time::SystemTime::now()
+            .duration_since(std::time::SystemTime::UNIX_EPOCH - std::time::Duration::from_secs(60))
+            .unwrap()
+            .as_secs() as u32,
+        bits: rng.gen(),
+        nonce: 0,
+    };
+    let start_time = Instant::now();
+    let mut hashes: u64 = 0;
+    let duration = Duration::from_secs(duration_secs);
+    let mut miner = Miner::new(0);
+    miner.new_target(vec![0_u8; 32]);
+    miner.header = Some(header);
+
+    while start_time.elapsed() < duration {
+        let _ = miner.next_share();
+        hashes += 1;
+    }
```

Comment:
why are we setting the target as 0?

doesn't this make it basically impossible to find a share?
```
difficulty = genesis_target / target = genesis_target / 0 = ∞
```

we are bypassing this restriction because we are ignoring the error returned by `miner.next_share()`, but that makes the code more confusing and counter-intuitive than it needs to be

but in order to improve readability and make a more clear usage of the API provided by `miner.next_share()`, I would refactor this snippet into something like:

```suggestion
    // set the max possible target, so that every share is valid
    let max_target = "ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff";
    let target = hex::decode(max_target).expect("Decoding failed");
    miner.new_target(target);
    miner.header = Some(header);

    while start_time.elapsed() < duration {
        miner.next_share().expect("every share is valid under max_target");
        hashes += 1;
    }
```

the downside for this approach is that now `next_share` will spam the console:
```
2024-05-03T20:59:11.630062Z  INFO mining_device: Found share with nonce: 0, for target: Some(0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff), with hash: 0x03986db2375a8106218561993e202c09b8ed969157417664ccba54c1ef8c5e6e
2024-05-03T20:59:11.630065Z  INFO mining_device: Found share with nonce: 0, for target: Some(0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff), with hash: 0x03986db2375a8106218561993e202c09b8ed969157417664ccba54c1ef8c5e6e
2024-05-03T20:59:11.630082Z  INFO mining_device: Found share with nonce: 0, for target: Some(0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff), with hash: 0x03986db2375a8106218561993e202c09b8ed969157417664ccba54c1ef8c5e6e
...
```

so `next_share` could probably have a filter for [the line that prints this](https://github.com/Fi3/stratum-1/blob/e4ad646c344d7ea124fbf651c575e4b956940bc3/roles/test-utils/mining-device/src/main.rs#L603):
```diff
-        if hash < *self.target.as_ref().ok_or(())? {
-            info!(
-                "Found share with nonce: {}, for target: {:?}, with hash: {:?}",
-                header.nonce, self.target, hash,
-            );
+        let target = *self.target.as_ref().ok_or(())?;
+        let max_target_hashrate_measurement =
+            Uint256::from_be_bytes(vec![255_u8; 32].try_into().unwrap());
+        if hash < target {
+            if target < max_target_hashrate_measurement {
+                info!(
+                    "Found share with nonce: {}, for target: {:?}, with hash: {:?}",
+                    header.nonce, self.target, hash,
+                );
+            }

```

Reply:
we don't car if share are valid or not we just want to execute the dsha function in the time interval and count how many time we are doing that so we know the hashrate of the machine that is running this cpuminer.

Reply:
the majior issue with the above code is that next_share return an Result (for sake of simplicity) instead of a custom type like enum ValidShare | InvalidShare

Reply:
i'm going do push something next week

Reply:
ok

it took me some time but I ended up understanding everything is going on

the suggested change is just a small cosmetic change to improve code readability, but not essential

feel free to ignore this and merge if you wish